suricata/update/main.py

```diff
@@ -424,15 +428,35 @@ def manage_classification(suriconf, files):
 def handle_dataset_files(rule, dep_files):
     if not rule.enabled:
         return
-    load_attr = [el.strip() for el in rule.dataset.split(",") if "load" in el][0]
-    dataset_fname = os.path.basename(load_attr.split(" ")[1])
-    filename = [fname for fname, content in dep_files.items() if fname == dataset_fname]
-    if filename:
-        logger.debug("Copying dataset file %s to output directory" % dataset_fname)
-        with open(os.path.join(config.get_output_dir(), dataset_fname), "w+") as fp:
-            fp.write(dep_files[dataset_fname].decode("utf-8"))
+
+    dataset_load = [el.strip() for el in rule.dataset.split(",") if el.startswith("load")]
+    if not dataset_load:
+        # No dataset load found.
+        return
+    dataset_filename = dataset_load[0].split(maxsplit=1)[1].strip()
+
+    # Get the directory name the rule is from.
+    prefix = os.path.dirname(rule.group)
+
+    # Construct the source filename.
+    source_filename = "{}/{}".format(prefix, dataset_filename)
+
+    if source_filename in dep_files:
+        content_hash = hashlib.md5(dep_files[source_filename]).hexdigest()
+        new_rule = re.sub("(dataset.*?load\s+){}".format(dataset_filename), "\g<1>datasets/{}".format(content_hash), rule.format())
+        dest_filename = os.path.join(config.get_output_dir(), "datasets", content_hash)
+        dest_dir = os.path.dirname(dest_filename)
+        logger.debug("Copying dataset file {} to {}".format(dataset_filename, dest_filename))
+        try:
+            os.makedirs(dest_dir, exist_ok=True)
+        except Exception as err:
+            logger.error("Failed to create directory {}: {}".format(dest_dir, err))
+            return
+        with open(dest_filename, "w") as fp:
+            fp.write(dep_files[source_filename].decode("utf-8"))
+        return new_rule
     else:
-        logger.error("Dataset file %s was not found" % dataset_fname)
+        logger.error("Dataset file '{}' was not found".format(dataset_filename))
```

Comment:
Should we instead log `source_filename` here to know where we are trying to load the file from and failing?

Reply:
No, I don't think so.  The `source_filename` will be something like `rules/my-dataset.txt`, and if you grep this in the ruleset it will never be found, as its in the rule as just `my-dataset.txt`. However, it probably makes sense to log the SID and MSG to one case more easily find out what rule to look for.